pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,599 @@
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util"
+	"k8s.io/kubernetes/pkg/volume/util/volumehelper"
+)
+
+// ProbeVolumePlugins is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&storageosPlugin{nil}}
+}
+
+type storageosPlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &storageosPlugin{}
+var _ volume.PersistentVolumePlugin = &storageosPlugin{}
+var _ volume.DeletableVolumePlugin = &storageosPlugin{}
+var _ volume.ProvisionableVolumePlugin = &storageosPlugin{}
+
+const (
+	storageosPluginName = "kubernetes.io/storageos"
+	storageosDevicePath = "/var/lib/storageos/volumes"
+	defaultAPIAddress   = "tcp://localhost:5705"
+	defaultAPIUser      = "storageos"
+	defaultAPIPassword  = "storageos"
+	defaultAPIVersion   = "1"
+	defaultFSType       = "ext4"
+	defaultNamespace    = "default"
+)
+
+func getPath(uid types.UID, namespace string, volName string, pvName string, host volume.VolumeHost) string {
+	if len(namespace) != 0 && len(volName) != 0 && strings.Count(volName, ".") == 0 {
+		return host.GetPodVolumeDir(uid, kstrings.EscapeQualifiedNameForDisk(storageosPluginName), namespace+"."+volName)
+	}
+	return host.GetPodVolumeDir(uid, kstrings.EscapeQualifiedNameForDisk(storageosPluginName), pvName)
+}
+
+func (plugin *storageosPlugin) Init(host volume.VolumeHost) error {
+	plugin.host = host
+	return nil
+}
+
+func (plugin *storageosPlugin) GetPluginName() string {
+	return storageosPluginName
+}
+
+func (plugin *storageosPlugin) GetVolumeName(spec *volume.Spec) (string, error) {
+	volumeSource, _, err := getVolumeSource(spec)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s/%s", volumeSource.VolumeNamespace, volumeSource.VolumeName), nil
+}
+
+func (plugin *storageosPlugin) CanSupport(spec *volume.Spec) bool {
+	return (spec.PersistentVolume != nil && spec.PersistentVolume.Spec.StorageOS != nil) ||
+		(spec.Volume != nil && spec.Volume.StorageOS != nil)
+}
+
+func (plugin *storageosPlugin) RequiresRemount() bool {
+	return false
+}
+
+func (plugin *storageosPlugin) GetAccessModes() []v1.PersistentVolumeAccessMode {
+	return []v1.PersistentVolumeAccessMode{
+		v1.ReadWriteOnce,
+		v1.ReadOnlyMany,
+	}
+}
+
+func (plugin *storageosPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volume.VolumeOptions) (volume.Mounter, error) {
+	return plugin.newMounterInternal(spec, pod, &storageosUtil{}, plugin.host.GetMounter())
+}
+
+func (plugin *storageosPlugin) newMounterInternal(spec *volume.Spec, pod *v1.Pod, manager storageosManager, mounter mount.Interface) (volume.Mounter, error) {
+	source, readOnly, err := getVolumeSource(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageosMounter{
+		storageos: &storageos{
+			podUID:          pod.UID,
+			namespace:       pod.Namespace,
+			pvName:          spec.Name(),
+			volName:         source.VolumeName,
+			volNamespace:    source.VolumeNamespace,
```

Comment:
Thanks - looking at this now + will complete the exception.  Also missed VisitPodSecretNames().